src/planner/binder/query_node/bind_setop_node.cpp

```diff
@@ -71,47 +95,166 @@ unique_ptr<BoundQueryNode> Binder::BindNode(SetOperationNode &statement) {
-		                         alias_map, expression_map, statement.left->GetSelectList().size());
-		BindModifiers(order_binder, statement, *result);
-	}
-
 	result->names = result->left->names;
 
 	// move the correlated expressions from the child binders to this binder
 	MoveCorrelatedExpressions(*result->left_binder);
 	MoveCorrelatedExpressions(*result->right_binder);
 
 	// now both sides have been bound we can resolve types
-	if (result->left->types.size() != result->right->types.size()) {
+	if (result->setop_type != SetOperationType::UNION_BY_NAME &&
+	    result->left->types.size() != result->right->types.size()) {
 		throw BinderException("Set operations can only apply to expressions with the "
 		                      "same number of result columns");
 	}
 
-	// figure out the types of the setop result by picking the max of both
-	for (idx_t i = 0; i < result->left->types.size(); i++) {
-		auto result_type = LogicalType::MaxLogicalType(result->left->types[i], result->right->types[i]);
-		if (!can_contain_nulls) {
-			if (ExpressionBinder::ContainsNullType(result_type)) {
-				result_type = ExpressionBinder::ExchangeNullType(result_type);
+	if (result->setop_type == SetOperationType::UNION_BY_NAME) {
+		case_insensitive_map_t<idx_t> left_names_map;
+		case_insensitive_map_t<idx_t> right_names_map;
+
+		BoundQueryNode *left_node = result->left.get();
+		BoundQueryNode *right_node = result->right.get();
+
+		for (idx_t i = 0; i < left_node->names.size(); ++i) {
+			if (left_names_map.find(left_node->names[i]) != left_names_map.end()) {
+				throw BinderException("UNION BY NAME operation doesn't support same name in SELECT list");
 			}
+			left_names_map[left_node->names[i]] = i;
+		}
+
+		for (idx_t i = 0; i < right_node->names.size(); ++i) {
+			if (right_names_map.find(right_node->names[i]) != right_names_map.end()) {
+				throw BinderException("UNION BY NAME operation doesn't support same name in SELECT list");
+			}
+			if (left_names_map.find(right_node->names[i]) == left_names_map.end()) {
+				result->names.push_back(right_node->names[i]);
+			}
+			right_names_map[right_node->names[i]] = i;
+		}
+
+		idx_t new_size = result->names.size();
+		bool need_reorder = false;
+		vector<idx_t> left_reorder_idx(left_node->names.size());
+		vector<idx_t> right_reorder_idx(right_node->names.size());
+
+		for (idx_t i = 0; i < new_size; ++i) {
+			auto left_index = left_names_map.find(result->names[i]);
+			auto right_index = right_names_map.find(result->names[i]);
+			bool left_exist = left_index != left_names_map.end();
+			bool right_exist = right_index != right_names_map.end();
+			LogicalType result_type;
+			if (left_exist && right_exist) {
+				result_type = LogicalType::MaxLogicalType(left_node->types[left_index->second],
+				                                          right_node->types[right_index->second]);
+				need_reorder = need_reorder || left_index->second != i || right_index->second != i;
```

Comment:
Predication here is less readable and not necessary, can we just change this into an `if` statement?